File: csharp/dnd-character/Iteration1/DndCharacter.cs
using System;
using System.Collections.Generic;
using System.Linq;

public class DndCharacter
{
    private DndCharacter() { }
    
    public int Strength { get; private set; }
    public int Dexterity { get; private set; }
    public int Constitution { get; private set; }
    public int Intelligence { get; private set; }
    public int Wisdom { get; private set; }
    public int Charisma { get; private set; }
    
    public int Hitpoints => 10 + Modifier(Constitution);

    public static int Modifier(int score) =>
        (int)Math.Floor((score - 10) / 2.0);
    
    public static int Ability() =>
        SixSidedDie
            .GetDiceRolls(4)
            .OrderByDescending(score => score)
            .Take(3)
            .Sum();

    public static DndCharacter Generate() =>
        new DndCharacter()
        {
            Strength = Ability(),
            Dexterity = Ability(),
            Constitution = Ability(),
            Intelligence = Ability(),
            Wisdom = Ability(),
            Charisma = Ability()
        };
}

public static class SixSidedDie
{
    public static IEnumerable<int> GetDiceRolls(int number) =>
        GetDiceRolls().Take(number);
    
    public static IEnumerable<int> GetDiceRolls()
    {
        var random = new Random();

        while (true)
            yield return random.Next(1, 7);
    }
}

File: c/dnd-character/Iteration1/dnd_character.c
#include <stdlib.h>
#include <math.h>

#include "dnd_character.h"

int ability(void) {
    int score = 0;
    int min = 6;

    for (int i = 0; i < 4; i++) {
        int val = (rand() % 6) + 1;
        score += val;

        if (val < min) {
            min = val;
        }
    }

    score -= min;

    return score;
}

int modifier(int score) {
    return floor((score - 10) / 2.0);
}

dnd_character_t make_dnd_character(void) {
    int constitution = ability();
    
    return (dnd_character_t) {
        .strength = ability(),
        .dexterity = ability(),
        .constitution = constitution,
        .intelligence = ability(),
        .wisdom = ability(),
        .charisma = ability(),
        .hitpoints = 10 + modifier(constitution)
    };
}
